File: bot/services/logging.py
"""Send mod and ticket logs to configured channel."""
import discord
from bot.database import queries as db
from bot.utils.embeds import base_embed


async def send_log(
    guild: discord.Guild,
    event_type: str,
    *,
    channel_id: str | None = None,
    user: discord.User | None = None,
    ticket_type: str | None = None,
    ticket_number: int | None = None,
    closed_by: discord.User | None = None,
    reopened_by: discord.User | None = None,
    escalated_by: discord.User | None = None,
) -> None:
    config = await db.guild_get_or_create(str(guild.id))
    log_channel_id = config.get("log_channel_id") or None
    if not log_channel_id:
        return
    channel = guild.get_channel(int(log_channel_id))
    if not channel or not isinstance(channel, discord.TextChannel):
        return
    embed = base_embed(timestamp=True)
    if event_type == "ticket_created" and channel_id and user and ticket_type is not None and ticket_number is not None:
        embed.title = "Ticket Created"
        embed.color = 0x57F287
        embed.add_field(name="Channel", value=f"<#{channel_id}>", inline=True)
        embed.add_field(name="User", value=f"{user.tag} ({user.id})", inline=True)
        embed.add_field(name="Type", value=ticket_type, inline=True)
        embed.add_field(name="Ticket #", value=str(ticket_number), inline=False)
    elif event_type == "ticket_closed" and channel_id and closed_by:
        embed.title = "Ticket Closed"
        embed.color = 0xFEE75C
        embed.add_field(name="Channel", value=f"<#{channel_id}>", inline=True)
        embed.add_field(name="Closed by", value=closed_by.tag, inline=True)
        embed.add_field(name="Ticket #", value=str(ticket_number or "—"), inline=True)
    elif event_type == "ticket_reopened" and channel_id and reopened_by:
        embed.title = "Ticket Reopened"
        embed.color = 0x57F287
        embed.add_field(name="Channel", value=f"<#{channel_id}>", inline=True)
        embed.add_field(name="Reopened by", value=reopened_by.tag, inline=True)
    elif event_type == "ticket_escalated" and channel_id and escalated_by:
        embed.title = "Ticket Escalated"
        embed.color = 0xEB459E
        embed.add_field(name="Channel", value=f"<#{channel_id}>", inline=True)
        embed.add_field(name="Escalated by", value=escalated_by.tag, inline=True)
    else:
        return
    await channel.send(embed=embed)

File: src/services/logging.ts
import type { Guild, TextChannel, User } from 'discord.js';
import { guildQueries } from '../database/queries.js';
import { baseEmbed } from '../utils/embeds.js';

export type LogEvent =
  | { type: 'ticket_created'; guild: Guild; channelId: string; user: User; ticketType: string; ticketNumber: number }
  | { type: 'ticket_closed'; guild: Guild; channelId: string; closedBy: User; ticketNumber?: number }
  | { type: 'ticket_reopened'; guild: Guild; channelId: string; reopenedBy: User }
  | { type: 'ticket_escalated'; guild: Guild; channelId: string; escalatedBy: User }
  | { type: 'mod_action'; guild: Guild; action: string; target: string; moderator: User; reason?: string };

export async function sendLog(guild: Guild, event: LogEvent): Promise<void> {
  const config = await guildQueries.getOrCreate(guild.id);
  const logChannelId = config.logChannelId ?? process.env.LOG_CHANNEL_ID;
  if (!logChannelId) return;

  const channel = await guild.channels.fetch(logChannelId).catch(() => null);
  if (!channel?.isTextBased() || channel.isDMBased()) return;

  const embed = baseEmbed({ timestamp: true });
  switch (event.type) {
    case 'ticket_created':
      embed
        .setTitle('Ticket Created')
        .setColor(0x57f287)
        .addFields(
          { name: 'Channel', value: `<#${event.channelId}>`, inline: true },
          { name: 'User', value: `${event.user.tag} (${event.user.id})`, inline: true },
          { name: 'Type', value: event.ticketType, inline: true },
          { name: 'Ticket #', value: String(event.ticketNumber), inline: false }
        );
      break;
    case 'ticket_closed':
      embed
        .setTitle('Ticket Closed')
        .setColor(0xfee75c)
        .addFields(
          { name: 'Channel', value: `<#${event.channelId}>`, inline: true },
          { name: 'Closed by', value: `${event.closedBy.tag}`, inline: true },
          { name: 'Ticket #', value: event.ticketNumber != null ? String(event.ticketNumber) : '—', inline: true }
        );
      break;
    case 'ticket_reopened':
      embed
        .setTitle('Ticket Reopened')
        .setColor(0x57f287)
        .addFields(
          { name: 'Channel', value: `<#${event.channelId}>`, inline: true },
          { name: 'Reopened by', value: `${event.reopenedBy.tag}`, inline: true }
        );
      break;
    case 'ticket_escalated':
      embed
        .setTitle('Ticket Escalated')
        .setColor(0xeb459e)
        .addFields(
          { name: 'Channel', value: `<#${event.channelId}>`, inline: true },
          { name: 'Escalated by', value: `${event.escalatedBy.tag}`, inline: true }
        );
      break;
    case 'mod_action':
      embed
        .setTitle(`Moderation: ${event.action}`)
        .setColor(0xed4245)
        .addFields(
          { name: 'Target', value: event.target, inline: true },
          { name: 'Moderator', value: `${event.moderator.tag}`, inline: true },
          ...(event.reason ? [{ name: 'Reason', value: event.reason, inline: false }] : [])
        );
      break;
  }

  await (channel as TextChannel).send({ embeds: [embed] }).catch(() => {});
}
